Enhance check_compliance to handle has_compliance_policies

kbrock opened the pull request:
### Before

- Not all classes implemented `has_compliance_policies?`
- Not all classes implemented `compliance_policies`
- api had to use one or the other to check if a model had policies assigned and therefore supported compliance.

### After

- All classes implement both methods
- supports uses `has_compliance_policies?` to properly answer supports question
- hosts no longer implements has_compliance_policies since the default works fine
- api no longer is concerned with these details

Fryguy commented on app/models/mixins/compliance_mixin.rb:
```diff
@@ -34,9 +36,15 @@ def scan_and_check_compliance_queue
   end
+
+  def has_compliance_policies?
+    !compliance_policies.blank?
```
```suggestion
    compliance_policies.present?
```

Fryguy commented on app/models/host.rb:
```diff
@@ -1693,11 +1693,6 @@ def self.host_create_os_types
-  def has_compliance_policies?
-    _, plist = MiqPolicy.get_policies_for_target(self, "compliance", "host_compliance_check")
-    !plist.blank?
-  end
```
Why was this removed?  I expected it to be similar to the others where the method was changed.

Fryguy replied:
Oh nvm - I didn't realize there was a common base method below.

Fryguy commented on app/models/mixins/compliance_mixin.rb:
```diff
@@ -9,7 +9,9 @@ module ComplianceMixin
 
-    supports :check_compliance
+    supports :check_compliance do
+      unsupported_reason_add(:check_compliance, "No compliance policies assigned") unless has_compliance_policies?
```
Do we have a way to i18n the text, or is that done later?

kbrock replied:
https://github.com/kbrock/manageiq/blob/compliance_policies/app/models/mixins/compliance_mixin.rb#L47-L49

kbrock replied:
There hasn't been i18n. I can add _ here

agrare commented on app/models/mixins/compliance_mixin.rb:
```diff
@@ -34,9 +36,15 @@ def scan_and_check_compliance_queue
+  # name is sometimes {target_class.gsub(/_/,'')}_compliance_check
   def compliance_policies
     target_class = self.class.base_model.name.downcase
     _, plist = MiqPolicy.get_policies_for_target(self, "compliance", "#{target_class}_compliance_check")
```
If we pull `"#{target_class}_compliance_check"` out into a method it looks like rather than overriding this "whole" method (it isn't long but still) you could just override `def compliance_check_policy_name`

Fryguy commented on app/models/mixins/compliance_mixin.rb:
```diff
@@ -34,9 +36,15 @@ def scan_and_check_compliance_queue
     Compliance.scan_and_check_compliance_queue(self)
   end
 
+  # a few of these need to override
+  # name is sometimes {target_class.gsub(/_/,'')}_compliance_check
   def compliance_policies
     target_class = self.class.base_model.name.downcase
     _, plist = MiqPolicy.get_policies_for_target(self, "compliance", "#{target_class}_compliance_check")
     plist
   end
```
Instead of overriding the entire method, maybe it's cleaner to create a method justf for the policy name, and then have the subclasses only override _that_.

```ruby
def compliance_policy_name
  "#{self.class.base_model.name.downcase}_compliance_check"
end

  def compliance_policies
    _, plist = MiqPolicy.get_policies_for_target(self, "compliance", compliance_policy_name)
    plist
  end
```

kbrock commented on app/models/compliance.rb:
```diff
@@ -66,13 +66,10 @@ def self.check_compliance(target, _inputs = {})
                                                                                          :number     => id}
       end
     end
-    target_class = target.class.base_model.name.downcase
-    target_class = "vm" if target_class.match("template")
```
do we want to use `compliance_policy_name` here?
That fixes the template => vm swap
but it will also change physical server and container images change.

Which I do not know if we want to deal with this change.

Fryguy replied:
I think that makes sense, yes...Might as well put the "override" where it belongs.

kbrock replied:
so it looks like there is only one override (vm and template) which is totally valid.
The others were unnecessarily overridden. (the difference between `underscore` and `downcase`)

kbrock commented on app/models/compliance.rb:
```diff
@@ -66,13 +66,10 @@ def self.check_compliance(target, _inputs = {})
-    target_class = "vm" if target_class.match("template")
-
-    unless target.respond_to?(:compliances)
+    unless target.supports?(:compliances)
```
I am reacting to the responds_to (adam has rubbed off on me here)

but the old way asking if the relation is defined is probably fine.
looking for suggestions on this one.

Fryguy replied:
This feels like the wrong change, because you've changed the way supports is defined.  Before it was either supported or not supported.  Now, you've introduced "sometimes supported" via `has_compliance_policies?`.

I think this needs to change further such that the raise on the next line takes into account the `unsupported_reason`.  Either that or this should be `target.class.supports?(:compliances)`.

@agrare Keep me honest here :)

agrare replied:
Right if you just want to know if the class supports it (aka bypass the block) then you can call `target.class.supports?(:compliances)`

kbrock replied:
nit/edit: call `target.class.support?(:check_compliance)`

that was such a good idea. thnx

Fryguy commented on app/models/miq_event.rb:
```diff
@@ -145,6 +145,7 @@ def self.raise_evm_job_event(target, options = {}, inputs = {}, q_options = {})
       :suffix => nil
     )
 
+    # note: this varies a little from compliance policy name
```
suggestion to make the comment more searchable

```suggestion
    # NOTE: this varies a little from compliance_policy_name
```

kbrock replied:
decided to remove these and leave the event base name / event policy name derivation alone.

Probably the best fix is to change the policies around the physical server and container image rather than introducing this concept.

kbrock replied:
looks like one place used underscore and confused me. swapping it with downcase fixes most of the special cases

Fryguy commented on app/models/mixins/compliance_mixin.rb:
```diff
@@ -35,8 +39,15 @@ def scan_and_check_compliance_queue
   end
+
+  def compliance_policy_name
+    "#{self.class.base_model.name.downcase}"
```
```suggestion
    self.class.base_model.name.downcase
```

Fryguy commented on app/models/mixins/compliance_mixin.rb:
```diff
@@ -35,8 +37,15 @@ def scan_and_check_compliance_queue
     plist
   end
+
+  def compliance_policy_name
```
Originally I was thinking this would include the full name (i.e. `"#{compliance_policy_name}_compliance_check"`), which is why I suggested the method name `compliance_policy_name`.  However, I see that you're using it as more of a prefix since it is able to be reused elsewhere (which I like better), so I think this method should be instead be named `compliance_policy_prefix`

kbrock replied:
agreed. I had this as a full name and then I found all the places that did the name derivation and since we have other names that are based upon this, I took the tangent.

thanks. fixed

Fryguy commented on app/models/mixins/compliance_mixin.rb:
```diff
@@ -9,7 +9,9 @@ module ComplianceMixin
 
-    supports :check_compliance
+    supports :check_compliance do
+      unsupported_reason_add(:check_compliance, _("No compliance policies assigned")) unless has_compliance_policies?
```
Comment on i18n that might have gotten lost in the other outdated comment stream: https://github.com/ManageIQ/manageiq/pull/21533#discussion_r741215956

kbrock replied:
I went ahead and rolled this back per your second comment.

We probably want to handle the i18n when we are constructing the error messages later on.

kbrock commented:
@Fryguy I think I addressed all your comments. I'm pretty happy with this now.

- I only have one support. Thanks @agrare for the tip on that one.
- I have not addressed i18n
- the overrides for the compliances largely went away once I properly `s/underscore/downcase/`
- Added specs around compliance mixin (creating those specs were the most time consuming of this whole PR)
- I'm content with `event_towhat` but if you have another name or destination in mind, just give the word.

Fryguy commented on app/models/mixins/compliance_mixin.rb:
```diff
@@ -35,8 +37,16 @@ def scan_and_check_compliance_queue
     plist
   end
+
+  def event_towhat
```
I think `event` is not granular enough since this is mixed in.  e.g. In vm.rb, what's this "event" referring to (could be an miq_event, a native event, a compliance event, etc).

```suggestion
  def compliance_policy_model_name
```

Fryguy replied:
Or even better, since you don't have overrides anymore, this can live directly in the `compliance_policies` method.

kbrock replied:
Deleted

But the concept is real.

If you grep through the code for `class.base_model.name` and/or `match("template")` you will notice that this logic resides across a number of files, mostly centered around compliance, event processing, event names, and to some extent `towhat`. Tends to be called target_class.

you can find it in
- models/compliance.rb
- models/miq_policy.rb
- models/miq_alert.rb
- models/miq_event.rb

some of them are using underscore and others are using downcase (which has me concerned)
most of them remember to swap the name template over to vm

Fryguy commented on spec/support/examples_group/shared_examples_for_compliance_mixin.rb:
```diff
@@ -0,0 +1,30 @@
+shared_examples_for "MiqComplianceMixin" do
```
MiqComplianceMixin isn't an actual thing, so I'm concerned this might get confusing.  Think you can just make this:

```suggestion
shared_examples_for "ComplianceMixin" do
```

Fryguy commented on spec/support/examples_group/shared_examples_for_compliance_mixin.rb:
```diff
@@ -0,0 +1,30 @@
+shared_examples_for "MiqComplianceMixin" do
+  context "MiqComplianceMixin methods" do
```
```suggestion
  context "ComplianceMixin methods" do
```

Fryguy commented:
@kbrock Think I have a name for the method + a couple of super-minor comments.

kbrock commented:
PUSH: change description and removed compliance_policy_name / event_towhat  compliance_policy_prefix
